File: sorting/bubblesort/bubblesort.py
from commons import arraycommons

def sort(arr):
    swapped = False
    total_it = 0
    real_array_size = len(arr) - 1
    for i in range(real_array_size):
        for j in range(0, real_array_size-i):
            if arr[j] > arr[j + 1]:
                swapped = True
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
            total_it += 1
        if not swapped:
            break

    print(f"Loop ran for {total_it} iterations \n")
    return arr

array_size = int(input("Enter the size of array:"))
array = arraycommons.populateArray(array_size)

print("Before sorting:")
arraycommons.printArray(array)

array = sort(array)

print("Sorted array is:")
arraycommons.printArray(array)

File: sorting/bubblesort/bubblesort.c
#include <stdio.h>
#include <stdbool.h>
#include "../../commons/arraycommons.h"

int swap(int *prev, int *next) {
	int temp = *prev;
	*prev = *next;
	*next = temp;
}

int sort(int array[], int array_size) {
	bool swapped = false;
	int total_it = 0;
	int real_array_size = array_size - 1;
	for (int i = 0; i < real_array_size; i++) {
		for (int j = 0; j < real_array_size - i; j++) {
			if (array[j] > array[j + 1]) {
				swap(&array[j], &array[j + 1]);
				swapped = true;
			}
			total_it++;
		}
		if (swapped == false) {
			break;
		}
	}

	printf("Loop ran for %d iterations \n\n", total_it);
}

int main() {
	int array_size;

	printf("Enter the size of array:");
	scanf("%d", &array_size);

	int array[array_size];

	populateArray(array, array_size);

	printf("Before sorting: \n");
	printArray(array, array_size);

	sort(array, array_size);
	printf("Sorted array is: \n");
	printArray(array, array_size);
}